astropy/units/quantity_helper/function_helpers.py

```diff
@@ -572,6 +573,20 @@ def aslayoutarray_helper(a, dtype=None):
         a = _as_quantity(a).value
     return (a, dtype), {}, out_unit, None
 
+@function_helper(helps={np.frombuffer, np.fromfile, np.fromiter, np.fromstring})
+def fromrawdata_helper(*args, **kwargs):
+    return args, kwargs, UNIT_FROM_LIKE_ARG, None
+
+@function_helper
+def fromfunction(function, shape, *, dtype=float, **kwargs):
+    zero_arg = (0,) * len(shape)
+    try:
+        out_unit = function(*zero_arg).unit
+    except Exception:
+        out_unit = UNIT_FROM_LIKE_ARG
```

Comment:
note to reviewers: I initially thought of doing something similar for `fromiter` but changed my mind after realizing that consuming an iterator could be destructive in some cases (generators).

Reply:
I agree not to bother!